36 Add mark, unmark transactions, bugfix

GavinSin opened the pull request:
- [X]  This pull request is complete
- [ ]  This pull request is a work in progress

### This PR aims to implement:

- Add mark, unmark transaction command with index
- Add done status to `Transaction` class, and corresponding done icon in UI
- Add DG guide for `markTxn` and `unmarkTxn`
- Resolve bugfix for #202 

---

### Acceptance Criteria

- By default, a new transaction is unmark. There is no done icon for that new transaction.
- When a transaction is marked via `markTxn`:
  - A done icon appears beside the transaction's person name
  - The done status is saved to json file upon closing of app
- When a transaction is unmarked via `unmarkTxn`:
  - no done icon appears beside the transaction's person name
  - The done status is saved to json file upon closing of app
- If `markTxn` executes on done transaction, and if `unmarkTxn` executes on undone transaction, the transaction details remain the same.
- If the json storage file does not have `isDone` field, the transaction is loaded as undone, and save as undone transaction upon closing of app.
- If a user edits a person name, the person names in transaction list is updated (both in `CommonModel`, and `UI`)
- If a user add new transaction or if a user edits a transaction or if a user marks/unmarks a transaction, the filtered transaction list in the UI reuses the existing filter.
- By default, when app is started, the existing filter used is to show all transactions.
---

### Further Works

- The `totalBalance` UI overview should track only undone transaction. New issue: #215
- The UI should have filter option for user to choose `totalBalance` track either undone, or all transaction. Related issue: #181 
 

closes #36 , #180 , #159, #202

SeeYangZhi commented on src/main/java/spleetwaise/address/logic/commands/EditCommand.java:
```diff
@@ -113,9 +115,30 @@ public CommandResult execute() throws CommandException {
         return new CommandResult(String.format(MESSAGE_EDIT_PERSON_SUCCESS, Messages.format(editedPerson)));
     }
 
+    private void updatePersonsInTransactions(
```
Understand that this is Draft still. Might need to add rationale for PR reviewers to understand the need to have this method

GavinSin replied:
This method is for bugfix #202 when a user edits a person name, the matching person name/s in the transaction list should get updated as well.

SeeYangZhi reviewed (changes requested):
Yet to look through entire code, lacking Documentation. Might want to consider adding UG DG as a separate issue if it is not intended to be added now.

rollingpencil commented on src/main/java/spleetwaise/transaction/logic/commands/UpdateStatusCommand.java:
```diff
@@ -0,0 +1,65 @@
+
+        if (targetIndex.getZeroBased() >= lastShownList.size()) {
+            logger.log(Level.WARNING, "Invalid transaction index: {0}", targetIndex.getZeroBased());
+            throw new CommandException("The transaction index provided is invalid");
```
I think in this case can use the Messages.MESSAGE_INVALID_TRANSACTION_DISPLAYED_INDEX here

GavinSin replied:
I have replace the message with the one in `Messages` class.

rollingpencil commented on src/main/java/spleetwaise/transaction/logic/commands/UpdateStatusCommand.java:
```diff
@@ -0,0 +1,65 @@
+        Transaction updatedTxn = txnToUpdate.setStatus(getUpdatedStatus());
+
+        model.setTransaction(txnToUpdate, updatedTxn);
+        model.updateFilteredTransactionList(TransactionBookModel.PREDICATE_SHOW_ALL_TXNS);
```
This would reset the existing filter applied after the markCommand. Is it possible to keep the filter and use it to filter the list after the specified transaction has been marked?

GavinSin replied:
I have added `updateFilteredTransactionList()` method which receives empty parameter, indicating the model to reuse existing filter.

rollingpencil reviewed (changes requested):
Some minor changes requested but overall looks good.

GavinSin commented:
> Yet to look through entire code, lacking Documentation. Might want to consider adding UG DG as a separate issue if it is not intended to be added now.

I have added the DG guide for markTxn and unmarkTxn now.

SeeYangZhi commented:
> > Yet to look through entire code, lacking Documentation. Might want to consider adding UG DG as a separate issue if it is not intended to be added now.
> 
> I have added the DG guide for `markTxn` and `unmarkTxn` now.

Any plans to add UG? Or creating a separate issue and PR for that?

SeeYangZhi commented:
Should the command names be `markPaid ` and `markUnpaid` instead? My take is that we can consider deviating from the `xxxTxn` namings since it is more intuitive to understand as `markPaid` and `markUnpaid` ? Not sure if this will cause any other issues

GavinSin commented:
> > > Yet to look through entire code, lacking Documentation. Might want to consider adding UG DG as a separate issue if it is not intended to be added now.
> > 
> > 
> > I have added the DG guide for `markTxn` and `unmarkTxn` now.
> 
> Looks like UG is the one that is added. Any plans to add DG? Or create a separate issue and PR for that?

oops, you are right. The one I added is UG. As for DG, I am waiting for finalization of glossary terms (as discussed in tele group) and add together in this PR.

What I gonna add on DG:
- Put Glossary section in front of user stories and use cases section.
- Glossary terms includes what positive, negative amount, done and undone transaction means.
- Convert all the acceptance criteria that I have been tested manually in local system under section "Appendix: Instructions for manual testing"

SeeYangZhi commented on src/main/java/spleetwaise/transaction/logic/commands/EditCommand.java:
```diff
@@ -91,9 +91,10 @@ private static Transaction createEditedTransaction(
         Set<Category> categories = editTransactionDescriptor.getCategories().orElse(txnToEdit.getCategories());
 
-        return new Transaction(id, person, amount, description, date, categories);
+        return new Transaction(id, person, amount, description, date, categories).setStatus(status);
```
What is the consideration behind this line, i.e. status not included as the constructor but through the setStatus method?

GavinSin replied:
This is due to the assumption that all newly created transactions are **Undone**, thus to create a **Done** transaction, you have to first create a `Transaction` then `setStatus`. In new discussion, we agree that you can create new Done transaction straight away so this should be change in new PR alongside with `addTxn` command with status field.

SeeYangZhi reviewed (approved):
Mainly nitpicking on the use of setStatus instead of the constructor.

SeeYangZhi commented:
May want to consider moving the status in front of the person name? In the return string of Txns

rollingpencil reviewed (approved):
Good job on this PR! Nice usage of polymorphism for reducing repeated code.
